mkosi/distributions/debian.py

```diff
@@ -28,63 +26,97 @@ def kernel_image(name: str, architecture: str) -> Path:
-
-        # Pretend we're lxc so debootstrap skips its mknod check.
-        run_with_apivfs(state, cmdline, env=dict(container="lxc", DPKG_FORCE="unsafe-io"))
-
-        # systemd-boot won't boot unified kernel images generated without a BUILD_ID or VERSION_ID in
-        # /etc/os-release. Build one with the mtime of os-release if we don't find them.
-        with state.root.joinpath("etc/os-release").open("r+") as f:
-            os_release = f.read()
-            if "VERSION_ID" not in os_release and "BUILD_ID" not in os_release:
-                f.write(f"BUILD_ID=mkosi-{state.config.release}\n")
-
-        if not state.config.local_mirror:
-            cls._add_apt_auxiliary_repos(state, repos)
+    @staticmethod
+    def repositories(state: MkosiState) -> list[str]:
+        repos = ' '.join(("main", *state.config.repositories))
+
+        if state.config.local_mirror:
+            return [f"deb [trusted=yes] {state.config.local_mirror} {state.config.release} {repos}"]
+
+        main = f"deb {state.config.mirror} {state.config.release} {repos}"
+
+        if state.config.release in ("unstable", "sid"):
+            return [main]
+
+        updates = f"deb {state.config.mirror} {state.config.release}-updates {repos}"
+
+        # Security updates repos are never mirrored
+        if state.config.release in ("stretch", "buster"):
+            security = f"deb http://security.debian.org/debian-security/ {state.config.release}/updates {repos}"
         else:
-            # Add a single local offline repository, and then remove it after apt has ran
-            state.root.joinpath("etc/apt/sources.list.d/mirror.list").write_text(f"deb [trusted=yes] {state.config.local_mirror} {state.config.release} main\n")
+            security = f"deb https://security.debian.org/debian-security {state.config.release}-security {repos}"
 
-        install_skeleton_trees(state, False, late=True)
+        return [main, updates, security]
+
+    @classmethod
+    def install(cls, state: MkosiState) -> None:
+        # Instead of using debootstrap, we replicate its core functionality here. Because dpkg does not have
+        # an option to delay running pre-install maintainer scripts when it installs a package, it's
+        # impossible to use apt directly to bootstrap a Debian chroot since dpkg will try to run a maintainer
+        # script which depends on some basic tool to be available in the chroot from a deb which hasn't been
+        # unpacked yet, causing the script to fail. To avoid these issues, we have to extract all the
+        # essential debs first, and only then run the maintainer scripts for them.
+
+        # First, we set up merged usr.
+        mergedusrextra = {
+            "mipsel"  : ["lib32", "lib64"],
+            "ppc64"   : ["lib32", "lib64"],
+            "ppc64el" : ["lib64"],
+            "s390x"   : ["lib32"],
+            "i386"    : ["lib64", "libx32"],
+            "amd64"   : ["lib32", "lib64", "libx32"],
```

Comment:
There's a bunch more architectures that need to be taken care of, even just restricting to EFI stuff there's riscv and loongarch that should really be handled

Reply:
I copied these from the debootstrap function that sets up merged usr. It doesn't have any special handling for risc and loongarch, see https://salsa.debian.org/installer-team/debootstrap/-/blob/master/functions#L1369

Reply:
loongarch is there?

```
	    loongarch64*)
			link_dir="lib32 lib64" ;;

```